av/bindings prep

diagprov opened the pull request:
I'd like to merge this into develop, in order that we can use it for ongoing projects and it isn't languishing.
For release, this breaks ABI compatibility. According to semver that probably means a 2.0.0 release.
I'd like to get some other stuff in before that, like constant time AES that is also languishing on a branch I now need to rebase against this.

As a summary of what I've prepared in support of easier bindings:

 * An all mode that includes public and symmetric key E4 variants (and will in future include all other variants) but either no storage defined, or either of the file storages (memory is ignored on the basis we want something sane).
 * Definable storage sizes for the storages used in symkey+pubkey modes (quick win).
 * clang-tidy integration (make tidy will do what you think)
 * make install for the all mode.
 * shared objects for the all mode.
 * Storage API changes to do the following:
   - allow side-by-side compilation: now all storage apis are specific to the variant they implement.
   - allow a storage to return a pointer when doing things like requesting the c2 pubkey. However if it returns NULL there remain functions to copy it into local memory.
   - (removed) return capabilities for the storage to avoid confusion in all mode.
 * With the changes to all to allow file storages, integrate and run all tests in all mode, and include all mode in the QA scripts, along with desktop hardening options.
 * Provide a "non-embedded" script to ensure the library is hardened.

Yes that's quite a bit. I've tackled some quick issues as part of this, while preparing the code for easier use with bindings.

diagprov committed:
This commit adds objects common to all CONF types into a common file.

A STORE=none is also defined. Using this will currently break tests as
there is no logic to prevent test use. This is intended for when client
code will implement its own storage logic.

diagprov committed:
Enable public/private modes to be compiled together by abusing
attributes in horrible ways.

This may need testing on older GCC implementations to ensure we haven't
broken embedded toolchains.

diagprov committed:
Fix nasty hack so it works.

diagprov committed:
Quick comment.

diagprov committed:
Allow E4_CLAGS+="-DE4_TOPICS_MAX" or E4_DEVICES_MAX to work, see #8.

diagprov committed:
Load checks for changing storage sizes. This should resolve #8.

diagprov committed:
Integrate clang-tidy into the build process.

Resolves #16.

diagprov committed:
Disable tidy, since github actions do not yet install clang-tidy in
their images.

diagprov committed:
Remove any dependency in protect+unprotect functions on the storage
layer.

This requires the addition of new storage primitives, as well as their
use.

diagprov committed:
Remove __attribute__((alias)) logic entirely; replace with defines.

diagprov committed:
All-in-one build.

This commit makes it possible to build an "all-in-one" version of libe4
including pubkey and symkey code, subject to the following constraints:

 - Storage type must be none. It is the responsibility of calling
   applications to define a storage.
 - A shared object is producted as well as a static library.
 - Users who want optimizations should specify E4_CFLAGS=-O2 for
   example.
 - Tests do not yet run successfully and should probably not be tried.
 - Storage interface has gained a new function. This allows a storage to
   tell E4 what it can store. This allows for the storage of symkey and
   public key clients in the same storage (users are responsible for not
   mucking this up) or for storages to identify themselves as storing
   one or the other and preventing errors at runtime through detection.

diagprov committed:
Fix install logic.

diagprov committed:
Add a script that will compile with appropriate optimizations security
functionality for desktop use.

diagprov committed:
Add tidy to qa script. Slight phony tweaks. Add nonembedded build
script.

diagprov committed:
Break out make shared into its own rule because macOS.

diagprov committed:
Just because we can, undocumented macOS support.

diagprov committed:
And also mark funcs extern on other systems in mode all.

veorq committed:
header include-once fix

odeke-em commented on include/e4/e4.h:
```diff
@@ -132,14 +66,14 @@ typedef struct _e4storage e4storage;
 
  */
-int e4c_protect_message(uint8_t *ciphertext,
+/*int e4c_protect_message(uint8_t *ciphertext,
```
Let's delete code that we aren't using instead of leaving comments in.

diagprov replied:
This needs to stay here as documentation for the common function format that is implemented. I could delete it, but in pubkey mode and symkey mode this is the actual signature... it's just hidden because C is ugly.

odeke-em commented on include/e4/e4.h:
```diff
@@ -165,75 +99,39 @@ message.
-int e4c_reset_topics(e4storage *store);
+    #define e4c_protect_message e4c_pubkey_protect_message
+    #define e4c_unprotect_message e4c_pubkey_unprotect_message
+#endif
```
Please add an endif comment to indicate which ifdef has been completed so
```c
#endif // E4_MODE_PUBKEY
```

That aids in easy groking of the code base instead of scrolling up, in the case of large ifdef bodies.

odeke-em commented on include/e4/e4symkey.h:
```diff
@@ -0,0 +1,23 @@
+                          e4storage *storage,
+                          const uint32_t proto_opts);
+
+#endif
```
```c
#endif // E4_SYMKEY_H
```

odeke-em commented on include/e4/options.h:
```diff
@@ -0,0 +1,34 @@
+ * their keys. */
+#define E4_OPTION_IGNORE_MISSING_PUBKEY    0x02
+
+#endif
```
```c
#endif // E4_OPTIONS_H
```

odeke-em commented on include/e4/proto.h:
```diff
@@ -0,0 +1,49 @@
+/*
+ * Teserakt AG LIBE4 C Library
+ *
+ * Copyright 2018-2020 Teserakt AG, Lausanne, Switzerland
```
The year in which the code was written suffices, as that field is supposed to be `YYYY`.
But also it'll be very odd to have to update the end dates every year.

So `Copyright 2020`

diagprov replied:
...

odeke-em commented on mk/all/rules.mk:
```diff
@@ -0,0 +1,31 @@
+
```
Please remove this blank line.

odeke-em commented on mk/all/objects.mk:
```diff
@@ -0,0 +1,6 @@
+
```
Please remove these blank lines.

odeke-em commented on src/e4pkcclient.c:
```diff
@@ -165,11 +177,25 @@ int e4c_protect_message(uint8_t *cptr,
+    }
+    
+    if ((privkey = (uint8_t*)e4c_get_idseckey_cached(storage)) == NULL) {
+
```
Unnecessary blank line.

odeke-em commented on src/e4pkcclient.c:
```diff
@@ -165,11 +177,25 @@ int e4c_protect_message(uint8_t *cptr,
                  const unsigned char *private_key);
      */
+    if ((pubkey = (uint8_t*)e4c_get_idpubkey_cached(storage)) == NULL) {
+
```
Unnecessary blank line.

odeke-em commented on src/e4pkcclient.c:
```diff
@@ -206,6 +233,11 @@ int e4c_unprotect_message(uint8_t *mptr,
 #endif
     
+    storagecaps = e4c_get_storage_caps(storage);
+    if (!(storagecaps & E4_STORECAP_PUBKEY)) {
```
Let's make this clearer and actually for a more grokable comparison since the result of this bitwise & is an *int*
```c
if ((storagecaps & E4_STORECAP_PUBKEY) == 0) {
```

odeke-em commented on src/e4symclient.c:
```diff
@@ -49,12 +49,23 @@ int e4c_protect_message(uint8_t *cptr,
+    uint32_t storagecaps;
+    
+    storagecaps = e4c_get_storage_caps(storage);
+    if (!(storagecaps & E4_STORECAP_SYMKEY)) {
```
Ditto about bitwise & then comparison against 0

diagprov replied:
This has disappeared as a problem for the moment, but I disagree that it's unclear. The logical not operator is more than fine with int types, especially as we don't have bool types in any event.

odeke-em reviewed (commented):
Thank you @diagprov! For an initial quick review:
a) For #ifdefs I'd highly encourage the style of adding a comment at the end of the #endif documenting which variable was being switched on, so for example
```c
#ifdef FOO_BAR
...
...
#endif // FOO_BAR
```
This aids with grokability/readability and later on with easily jumping through definitions, when the
ifdef clause is large.

diagprov commented:
I'm closing this for now as more work needs to be done before this can be integrated.

diagprov committed:
Enable side-by-side storages. The "all" mode can now build and use both
pubkey and symkey storages alongside each other.

It is assumed that the correct storage is passed to the relevant
e4c_protect / e4c_unprotect implementation, and there is (currently) no
way to type check this.

diagprov committed:
Comment most include guards.

diagprov committed:
ALL config added to github actions.

diagprov committed:
slight tidy up

diagprov committed:
Version: make it easy to set version numbers by changing one makefile.

diagprov committed:
Static analysis: check write() return values.

diagprov commented:
and I'm done. Side by side storages now work all in one build.

diagprov commented:
Gonna hit merge on this so JP can use the bindings from the develop branch, and since develop is for develop anyway :)
